### PR TITLE
support for callback function for SFTP::put function - in order to pipe ...

### PR DESCRIPTION
**Pull request by andrey012**
...data directly to remote server without putting it into file or keeping in memory. This can be useful particularly for dumping big databases directly to remote server.

### CODE REVIEW

File: phpseclib/Net/SFTP.php [hunk 1]
```diff
@@ -1759,6 +1764,10 @@ function rmdir($dir)
      *
      * If $data is a resource then it'll be used as a resource instead.
      *
+     *
+     * Setting $mode to self::SOURCE_CALLBACK will use $data as callback function, which gets only one parameter -- number
```
Comment (bantu) [inline phpseclib/Net/SFTP.php]: NET_SFTP_CALLBACK ?

**Comment by bantu** [comment]
@terrafrost The put() function already had support for an $callback paramter (via #660). Now $data can be a callback too. What's going on here?

File: phpseclib/Net/SFTP.php (hunk 1 above)
Comment (andrey012) [inline phpseclib/Net/SFTP.php]: Opps, sorry, my fault, thank you for report.
Fixed in another PR, commit 0cc6125f8795f577aa9e01535c1a58545e290866

AG

Andreas Fischer wrote:

**Comment by bantu** [comment]
@andrey012 The tests you added seem to be failing too, now.

**Comment by andrey012** [comment]
I'm not @terrafrost, but since I've recently added this functionality -- I'll answer.

Callback, that you mention is different thing.
1. in SCP, not SFTP.
2. it is used to report progress.

The callback, that I added to SFTP is used to pipe data without keeping it in memory (I used it to pipe from stdout of mysqldump launched via proc_open).

AG

Andreas Fischer wrote:

**Comment by bantu** [comment]
@andrey012 I am talking about SFTP. Something is just not right, here.

$callback function parameter
https://github.com/phpseclib/phpseclib/blob/1.0/phpseclib/Net/SFTP.php#L1801
$callback thrown away
https://github.com/phpseclib/phpseclib/blob/1.0/phpseclib/Net/SFTP.php#L1849
$callback taken from $data
https://github.com/phpseclib/phpseclib/blob/1.0/phpseclib/Net/SFTP.php#L1855

call to $callback
https://github.com/phpseclib/phpseclib/blob/1.0/phpseclib/Net/SFTP.php#L1899
anothr call to $callback
https://github.com/phpseclib/phpseclib/blob/1.0/phpseclib/Net/SFTP.php#L1916

**Comment by andrey012** [comment]
Andreas, thank you for report. Actually, as I can see - there were two PRs adding callbacks, one added callback to report progress (David Anderson in December 2014) , another was mine. But at the time I have made my change - David's one was not yet merged.
I've tried to sort this out in e7719f765c22b4cdce55a8ba806df7664bce1fb6, my another PR.

AG
Andreas Fischer wrote:

**Pull request merged**